### DISCUSSION TITLE
Non reentrant iterator warning

### DISCUSSION THREAD

**Discussion by Divelix** [libktx/ktx#371] [Q&A] [answered]
Why do I have non reentrant iterator warning from Android Studio even when I import `ktx.collections.*`?
![Screenshot_9](https://user-images.githubusercontent.com/25386431/121777531-51f03c00-cb9b-11eb-8775-420588ad12d5.png)

**Comment by czyzby** [comment, marked as answer]
This is because LibGDX reuses iterators internally and might return the same instance on subsequent calls. It's basically an anti-pattern, which I'm guessing they implemented to limit the garbage collection on mobiles. See [this StackOverflow thread](https://stackoverflow.com/questions/50237364/what-does-using-non-reentrant-iterator-method-array-iterator-error-message).

**Reply by Divelix** [reply]
So is it safe for me to just ignore this warning then?

**Reply by czyzby** [reply]
@Divelix If you're not iterating over the same collection concurrently or within a nested loop, you should be fine. Just keep in mind that something like this could have unexpected results if the same iterator is used by each loop:

```kotlin
for ((enemy, enemyCount) in Data.matchHistory) {
  for ((enemy, enemyCount) in Data.matchHistory) {
    // ...
  }
}
```

For safety reasons alone, you might be better off using the maps from the standard library unless this is a performance-critical part of your application.
